### ISSUE TITLE
Docker 1.12 RC4 swarm mode - Doesn't seem to be a way limit only one replica per node

### ISSUE THREAD

**Issue by timhaak** [moby/moby#24649]
Looking at the available documentation and this #24397 issue the only constraints available are based either on a node label or engine label. 

There doesn't seem to be a way to add a constraint of only one replica running on a node. 

Previously it could be done on the container name. (affinity:container!=name) 

Is this something that isn't available yet?

If so is there a plan to add this in the future?

**Comment by thaJeztah** [comment]
I think you're looking for `--mode=global`; https://docs.docker.com/engine/reference/commandline/service_create/#/set-service-mode

**Comment by timhaak** [comment]
Thanks. That will help.

Though that will start something on every node.

The example I'm looking for is:

You have 10 nodes and want to start a db services with a replication factor of 3 and you want to ensure that the db's are never running on the same host.

The simple solution would be to just specify the nodes to install it on.

Just thought I should mention it as I imagine it's common use case, so might be a nice thing to add for the future.

Otherwise really liking the new swarm in 1.12 :) 

Thanks

**Comment by thaJeztah** [comment]
Ah! Yes, makes sense.

let me ask @aluzzardi :smile:

**Comment by thaJeztah** [comment]
also /cc @sfsmitcha

**Comment by yongtang** [comment]
I think this is about setting constraint on the Tasks of Swarm so that each task should have different nodeID.

In Mesos/Marathon that could be achieved with `"constraints": [["hostname", "UNIQUE"]]` I believe.

**Comment by bargenson** [comment]
Any news on the ability to use affinities to avoid two instances to be on the same node when possible?

**Comment by Richard-Mathie** [comment]
A quick work around is that you set the containers in `--mode global` and then set a node constraint, i.e. `--constraint node.role==manager` will start only one container on each manager.

similarly you can set a node label and provision the containers that way `--constraint node.labels.mylabel==myvalue`. then you have to set `docker node update --add-label mylabel=myvalue` on each node you want on each node you want the service to run.

Unfortunately this removes the automatic provisioning of n instances on n unique nodes form a pool of m nodes. However you can see docker cant really have both a unique node affinity and a replication strategy, as the constraints are ambiguous when the number of nodes is less than the replication factor.

**Comment by timhaak** [comment]
@Richard-Mathie 

In the previous version of swarm it would then not reach the replication factor.

Ie you have 3 nodes and added a filter saying no other container with my name and a replication factor of 5.

It would then start 5 nodes.

Though I like the work arround as a temporary solution.

**Comment by joeljacobs** [comment]
This is a real problem for anything that requires HA failover. I'm setting up a Database HA service that has a master streaming to slave relationship. If both the master and slave happen to be on the same node, a node failure can completely bring down the service and it's data.

**Comment by rolele** [comment]
I am having the same problem with kafka.
I have 4 nodes but I only want kafka on 3 nodes so --global is not suitable.
```
[root@manager0 vagrant]# docker service ls
ID            NAME       REPLICAS  IMAGE                   COMMAND
9hoouijuogwj  docker-ui  global    uifd/ui-for-docker
ckyu7nyccn0f  kafka      3/3       wurstmeister/kafka
drwhj119pgu7  zookeeper  1/1       wurstmeister/zookeeper
[root@manager0 vagrant]# docker service ps kafka
ID                         NAME         IMAGE               NODE      DESIRED STATE  CURRENT STATE          ERROR
44zxiag7bvim999qhvtjhcmfi  kafka.1      wurstmeister/kafka  worker0   Running        Running 7 minutes ago
1yxvigcnvpodt9kq6koh4a2n1  kafka.2      wurstmeister/kafka  manager1  Running        Running 7 minutes ago
5u6qj9r7qx392cyniug7mpuk5  kafka.3      wurstmeister/kafka  worker0   Running        Running 7 minutes ago
````

has you can see kafka has been started 2 times on worker0.
When creating the service with ansible I expose the port 9092
```
  --name kafka
  --replicas 3
  --network appnet
  -p 9092:9092
  -e KAFKA_ADVERTISED_HOST_NAME=kafka
  -e KAFKA_ADVERTISED_PORT=9092
  -e KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181
  -e KAFKA_CREATE_TOPICS="kafkapipeline:1:1"
  wurstmeister/kafka
```
both containers ports should colide and fail though but nothing happen
```
[root@worker0 vagrant]# docker ps
CONTAINER ID        IMAGE                       COMMAND             CREATED             STATUS              PORTS               NAMES
2e1406ec8bb0        wurstmeister/kafka:latest   "start-kafka.sh"    12 minutes ago      Up 12 minutes                           kafka.3.5u6qj9r7qx392cyniug7mpuk5
cf379bced5e1        wurstmeister/kafka:latest   "start-kafka.sh"    12 minutes ago      Up 12 minutes                           kafka.1.44zxiag7bvim999qhvtjhcmfi
30612520e350        uifd/ui-for-docker:latest   "/ui-for-docker"    15 minutes ago      Up 15 minutes       9000/tcp            docker-ui.0.0l2tawaoh1pmsl57ezjo8fp07
```
No port forwarding has been setup by docker.

**Comment by rolele** [comment]
The problem with the work around of @Richard-Mathie is that the service stay in: Allocated state

```
[root@manager0 vagrant]# docker service ps kafka
ID                         NAME       IMAGE               NODE      DESIRED STATE  CURRENT STATE             ERROR
eel2psv0g53esz1nzua86ognu  kafka      wurstmeister/kafka  manager1  Running        Running 19 minutes ago
8imbzx1tjqdjgnasymsnnu1lc   \_ kafka  wurstmeister/kafka  manager2  Running        Running 19 minutes ago
dtfsuvr2l9rwtdekrbhb17pb7   \_ kafka  wurstmeister/kafka  worker0   Running        Allocated 19 minutes ago
9qask1a5iu5ctmtc93ibj1c00   \_ kafka  wurstmeister/kafka  manager0  Running        Running 19 minutes ago
```

**Comment by Richard-Mathie** [comment]
Not with the latest build (1.13.0 dev) of docker. I believe....

**Comment by pascalandy** [comment]
+1. [My use case explain here](https://github.com/docker/docker/issues/28393#issue-189150910).

**Comment by aluzzardi** [comment]
Hey everyone,

I just answered very similar questions in https://github.com/docker/docker/issues/26259#issuecomment-260899732 :)

TL;DR summary:
- Try 1.13.0-rc1 please, you'll like it :)
- The scheduler does that for you by default now. No more than one container per machine if possible 
- Even the workaround that was explained here would work better not (no more tasks stuck in allocated state). But you don't need the workaround :)
- Please provide feedback! 1.13.0 is just around the corner and we can make it better for you if you speak up now

/cc @aaronlehmann

**Comment by yaakov-berkovitch** [comment]
Hi,
I'm running docker version 1.13.0-rc3, build 4d92237 (experimental version) and deploying using docker deploy <compose-file>, and it's not working for me. Several replicas are running on the same node.
In the compose file, I defined affinity based on the container for each service.
Can you please confirm that it should work on this version ?

Thanks.

**Comment by thaJeztah** [comment]
@yaakov-berkovitch swarm mode does not have `affinity:` filer like ["classic" swarm does](https://docs.docker.com/swarm/scheduler/filter/#use-an-affinity-filter). You can set _constraints_ to have a service only run on a node that has the specified labels, see [Specify service constraints (--constraint)](https://github.com/docker/docker/blob/v1.13.0-rc3/docs/reference/commandline/service_create.md#specify-service-constraints---constraint).

**Comment by yaakov-berkovitch** [comment]
Thanks for the reply. In my case I would like to run a "global" service running on each node and exactly only one instance.
Can you please advise me how to do it if I cannot write a constraints based on the service name ?

**Comment by thaJeztah** [comment]
> In my case I would like to run a "global" service running on each node and exactly only one instance.

@yaakov-berkovitch you can create a "global" service, as I mentioned above; https://github.com/docker/docker/issues/24649#issuecomment-232805150

**Comment by yaakov-berkovitch** [comment]
Thanks. It was also my understanding but I "fight" to find the exact syntax in the compose file to specify such policy. Finally found that there is new "deploy" section that allows me configuring the global mode.
:-)

**Comment by JnMik** [comment]
The closest thing I can to this is "Deployment strategy : High availability" on this page : 
https://docs.docker.com/docker-cloud/infrastructure/deployment-strategies/

But it seems to be only available on Docker cloud -_-

**Comment by thaJeztah** [comment]
@JnMik what version of docker are you running? See https://github.com/moby/moby/issues/24649#issuecomment-260901703

**Comment by JnMik** [comment]
Docker version 1.13.0, build 49bf474
But very often my containers are deployed on the same node.
(Even if the other nodes still have ressources available)

**Comment by thaJeztah** [comment]
@JnMik first of all, it may be worth updating to docker 17.03 (which is essentially what would've been docker 1.13.2) - the current 17.03 version is 17.03.2, so there's been 4 patch releases since 1.13 

If you have more information (possibly logs), and steps to reproduce, could you open a separate issue for this?

**Comment by JnMik** [comment]
I'll check it out with the latest version !
thx

**Comment by JnMik** [comment]
Seems to work so far! gj

**Comment by thaJeztah** [comment]
That's good to hear, thanks!

**Comment by thaJeztah** [comment]
Let me close this one, as it went stale, and new options were added since this was reported (see https://github.com/moby/moby/issues/26259)
